### DISCUSSION TITLE
RFC: Changing `em.find` methods interface in v6?

### DISCUSSION THREAD

**Discussion by B4nan** [mikro-orm/mikro-orm#3982] [Polls]
I am thinking about changing the `em.find` methods interface, moving the `where` parameter to the `FindOptions`:

```diff
-const res = await em.find(User, { name: 'foo' }, { populate: [...], orderBy: { ... } });
+const res = await em.find(User, { where: { name: 'foo' }, populate: [...], orderBy: { ... } });
```

This would affect all the `find` methods, including `findOne` and its variants. We would probably provide a codemod for the migration.

Since this is quite a breaking change (mainly for me, so many tests to rewrite! :trollface:), I'd like to see some confirmation. I am also adding one more option with a single `options` parameter, but I would rather not go that way, as it would complicate the inference most probably - and it would be the same interface as repository, so if that's the goal, you can always just use repository.

I would rather not maintain both variants (kinda like v4 did) as it complicates the code/docs quite a lot. And maybe this is not an issue at all, and not worth the hassle. 

So I'd like to hear some feedback!

**Comment by roderik** [comment]
I voted to keep it like v5, just because it is also a ton of work in our app to change them all and the benefit feels more cosmetic than anything

**Reply by B4nan** [reply]
> feels more cosmetic than anything

I'd argue the change would make the code generally more readable, I don't see that as something cosmetic. I'd like to target new users, of course this won't make life of existing users much better.

**Reply by B4nan** [reply]
I guess we could have a codemod for this to make the transition easier. I feel like writing one will save me a lot of time, there are more than 1000 occurrences in our tests.

**Reply by roderik** [reply]
or a small v5-compat package that adds the old signatures?

**Reply by B4nan** [reply]
I'd rather not, especially because both `where` and `options` are POJOs, there is no easy way to distinguish those (unless there is an explicit third `options` parameter).

**Comment by B4nan** [comment]
I will wait/hope for more votes, but few notes on how I interpret this so far:
- the general consensus is rather yes, let's do this
	- would be great to have some feedback on the "no" answers, as it feels it's more about the back compatibility
- nobody likes BCs, so a codemod would make a lot of sense, which resolves the BC issues to me
- as mentioned above, having the non-breaking variant is rather a no-go, due to the fact that both `where` and `options` are POJOs and there could be ambiguity 
	- so to me those votes go to the first option, if there would be a codemod, why staying on the old syntax?

**Comment by nsams** [comment]
Incompatible changes handled with a good codemod are no big issue.

**Comment by rubiin** [comment]
I  voted for keep as it is in v5 . Other improvement that are important should be addressed

**Comment by bakura10** [comment]
What is the status on this for v6? I have to admit that I’m being always caught by surprise on this one when using v5 😅

**Reply by B4nan** [reply]
I didn't want to do this in the end, as it felt really weird for anything else than `em.find`. Do you think it makes sense to add two new methods?

```ts
em.findByOptions(User, { ... }) // returns User[]
em.findOneByOptions(User, { ... }) // returns User | null
```

Another, maybe better idea was (since it makes most sense for `em.find`) to add `em.findAll()` and a `where` option to its options parameter (and do the same for the repository method). I think I will at least do that as its a simple change (edit: #4946).

--- 

I guess it's a bit too late for such a breaking change (I really wanted to go stable this week, although still finishing some last-minute changes), but for v7 we could add `em.findBy` which would have the second where parameter and keep `em.find` without it (and remove `em.findAll`). Still not sure if this all makes sense for `em.findOne`, but maybe it could also work as a universal method, with additional options that to trigger the same behaviour as `em.findOneOrFail`.

**Reply by bakura10** [reply]
I'm not a big fan of this findByOptions. It creates a discrepancy as you would also need to introduce a findByCursorByOptions and so on.

I'm not sure if this is possible, but isn't it possible to handle it in a backward-compatible way?

If you have `en.find(User, { where: ..., groupBy: ...} }`

You can probably detect the first hash contains a where _and_ other options, and rewrite this internally?

**Reply by B4nan** [reply]
Which will break as soon as someone defined a property called `where` on their entity. The findAll method seems like a better solution.
